src/reinforcement/mesh.py

```diff
@@ -0,0 +1,224 @@
+# This file covers all functions related to the mesh generation in gmsh and its conversion into separated xdmf files, which can then be used in dolfinx.
+
+import gmsh
+import numpy as np
+import meshio
+import dolfinx as dfx
+from mpi4py import MPI
+from numpy.typing import ArrayLike
+import math
+
+def create_concrete_slab(
+    point1: ArrayLike,
+    point2: ArrayLike,
+    n_x: int,
+    n_y: int,
+    margin: float,
+    s: float,
+    filename: str,
+    where="both",
+):
+    """
+    This function creates a 3D-quadrilateral mesh in the shape of a cuboid with reinforcement
+    rebars as line elements in a 3D space. Only a very general case with a uniform rebar-grid
+    at the top and/or bottom of the slab is possible.
+
+    Parameters
+    ----------
+    point1: ArrayLike
+        Starting point of the cuboid.
+    point2: ArrayLike
+        Endpoint of the cuboid.
+    n_x: int
+        Number of reinforcement bars in x-direction. 
+    n_y: int
+        Number of reinforcement bars in y-direction.
+    margin: float
+        Absolute value of the smallest distance between the outer edges of the concrete mesh
+        and a reinforcement element.
+    s: float
+        Maximal element size. It may be corrected according to n_x and n_y, such that
+        the reinforcement discretization is always held true and reinorcement&concrete
+        elements share the same nodes.
+    filename: str
+        Name of the msh file.
+    where: str
+        Position of the reinforecement bars. Valid options are "upper", "lower"
+        and "both".
+
+    Returns
+    -------
+
+    """
+
+    x0, y0, z0 = point1
+    l, w, h = point2
+
+    # initialize gmsh
+    gmsh.initialize()
+
+    # alias to facilitate code writing
+    mymesh = gmsh.model
+
+    # create the first point
+    point1 = mymesh.occ.addPoint(x0, y0, z0)
+    mymesh.occ.synchronize()
+
+    #  correct number of concrete elements (defined by s) in order to fit the coice of n_x and n_y
+    def num_elems(min_amount_elems, reinf_elems):
```

Comment:
Three functions are defined within `create_concrete_slab`. Could we reduce that? If it is important to have these functions, we can move them out of the function to improve readability and maybe name them `_function_name` to discourage use of them.